packages/core/src/routes/admin-user.ts

```diff
@@ -0,0 +1,43 @@
+import { userInfoSelectFields } from '@logto/schemas';
+import pick from 'lodash.pick';
+import { InvalidInputError } from 'slonik';
+import { object, string } from 'zod';
+
+import koaGuard from '@/middleware/koa-guard';
+import { findRolesByRoleName } from '@/queries/roles';
+import { findUserById, updateUserById } from '@/queries/user';
+
+import { AuthedRouter } from './types';
+
+export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
+  router.patch(
+    '/users/:userId/roles',
+    koaGuard({
+      params: object({ userId: string().min(1) }),
+      body: object({ roleNames: string().array().nullable() }),
+    }),
+    async (ctx, next) => {
+      const {
+        params: { userId },
+        body: { roleNames },
+      } = ctx.guard;
+
+      await findUserById(userId);
+
+      // Temp solution to validate the existence of input roleNames
+      if (roleNames?.length) {
+        const roles = await findRolesByRoleName(roleNames);
+        if (roles.length !== roleNames.length) {
+          const resourcesNotFound = roleNames.filter(
+            (rolesNames) => !roles.some(({ name }) => name === rolesNames)
+          );
+          throw new InvalidInputError(`role names (${resourcesNotFound.join(',')}) are not valid`);
```

Comment:
should be a `RequestError` with `data`?

Reply:
More like a DB foreign key violation error, like Slonik NotFound error. Think about adding an error mapper at the error handler middleware we have another issue that tracks the global error handler updates. I'll add a comment here just for reminding.